walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,80 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+
+        Collection<TimeRange> optional_attendees = new ArrayList<TimeRange>();
+        Collection<TimeRange> mandatory_attendees = new ArrayList<TimeRange>();
+        optional_attendees.add(TimeRange.WHOLE_DAY);
+        mandatory_attendees.add(TimeRange.WHOLE_DAY);
+
+        int minutesInDay = 24 * 60;
+        if (request.getDuration() > minutesInDay)
+            return new ArrayList<TimeRange>();
+        
+
+        Collection<String> allAttendees = new ArrayList<String>();
+        allAttendees.addAll(request.getAttendees());
+        allAttendees.addAll(request.getOptionalAttendees());
+        for (Event event : events) {
+            if (attendeeAvailability(request.getAttendees(), event))
+            {
+                mandatory_attendees = availableTimes(mandatory_attendees, event, request.getDuration());
+            }
+            if (attendeeAvailability(allAttendees, event))
+            {
+                optional_attendees = availableTimes(optional_attendees, event, request.getDuration());
+            }
+        }
```

Comment:
So e = number of events and p = number of total people between mandatory and optional. 

So we loop over all events. In the first if we check for attendee availability. This loops through all the attendees and gives true if any of the attendees have that meeting. Worst case it is always the last person in the list of attendees. There is the O(e*p). 

Then if the case is true (worst case it is always true) we call availableTimes. Thinking of the runtime of availableTimes is a bit confusing. It looks like it time it is called, it can potentially split one item into at most two items. The number of items in availabilities is O(e). Since we cycle through each item in the list, bringing it back to the parent function, we get O(e*(p+e)). The second if statement is the same so we get O(2*e(p+e)) but we don't care about constants so O(e*p + e^2). 

The best I found so far was O(e*p + e*log(e)). We don't know how log(e) compares to p so we can't drop either term. 

What if you first sorted the events by start time in O(e*log(e)). There should be a built in sorting mechanism for java.
Then looped through all events and people and created a list of all relevant events in O(e*p) in a way that maintained the sorting.
Then since all the event are sorted, merging overlapping events can be done in O(e).
Then you can get a list of all the times that aren't during events in O(e).

In total this becomes O(e*log(e) + e*p + e + e) but since e*log(e) > e we don't care about the e terms. This means it simplifies to O(e*log(e) + e*p).